chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        match self.shard_catchup_finalize_storage(shard_uid, &flat_head) {
+            Ok(_) => {
+                let task_status = FlatStorageReshardingTaskStatus::Successful { num_batches_done };
+                info!(target: "resharding", ?shard_uid, ?task_status, "flat storage shard catchup task finished");
+                // At this point we can trigger the reload of memtries.
+                self.scheduler.send(ReshardingRequest::MemtrieReload { shard_uid });
+                task_status
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, "flat storage shard catchup finalize failed!");
+                FlatStorageReshardingTaskStatus::Failed
+            }
+        }
+    }
+
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
+
+        let mut num_batches_done: usize = 0;
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
+                ?shard_uid,
+                ?flat_head_block_hash,
+                batch_id = ?num_batches_done)
+            .entered();
+            let chain_final_head = chain_store.final_head()?;
+            let mut merged_changes = FlatStateChanges::default();
+            let store = self.runtime.store().flat_store();
+            let mut store_update = store.store_update();
+
+            // Merge deltas from the next blocks until we reach chain final head.
+            for _ in 0..CATCH_UP_BLOCKS {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
+                // Stop if we reached chain final head.
+                if flat_head_block_hash == chain_final_head.last_block_hash {
+                    break;
+                }
+                flat_head_block_hash = chain_store.get_next_block_hash(&flat_head_block_hash)?;
+                if let Some(changes) = store
+                    .get_delta(shard_uid, flat_head_block_hash)
+                    .map_err(|err| Into::<StorageError>::into(err))?
+                {
+                    merged_changes.merge(changes);
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
```

Comment:
I would feel a bit better if this was using some flat storage API rather than reading directly from db.

Reply:
I don't think we can use flat storage API (if it exits) since the flat storage for this child doesn't yet exist in flat storage manager